### PR TITLE
Holograph Protocol Upgrade for Optimism (Bedrock including)

### PR DESCRIPTION
**Pull request by ACC01ADE**
## Describe Changes

Making necessary changes for supporting Optimism L2 in the Holograph Protocol.

- Optimism L1 Security Fee is not supported and calculated in LayerZeroModule
  - base gas price for L1 fee is set to 100 GWEI for now
- Also configuring minimum gas prices for networks

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] Code styles have been enforced
- [ ] All Hardhat tests are passing

### CODE REVIEW

File: deploy/10_layer_zero_module.ts [hunk 1]
```diff
@@ -109,13 +195,55 @@ const func: DeployFunction = async function (hre1: HardhatRuntimeEnvironment) {
 
   const holograph = await hre.ethers.getContract('Holograph', deployer);
 
+  const futureOptimismGasPriceOracleAddress = await genesisDeriveFutureAddress(
```
Comment (alexanderattar) [inline deploy/10_layer_zero_module.ts:202]: This needs to be deployed via Super Cold Storage right?

File: src/HolographOperator.sol [hunk 2]
```diff
@@ -516,7 +516,7 @@ contract HolographOperator is Admin, Initializable, HolographOperatorInterface {
   ) external payable {
     require(msg.sender == _bridge(), "HOLOGRAPH: bridge only call");
     CrossChainMessageInterface messagingModule = _messagingModule();
-    uint256 hlgFee = messagingModule.getHlgFee(toChain, gasLimit, gasPrice, bridgeOutPayload.length);
+    uint256 hlgFee = messagingModule.getHlgFee(toChain, gasLimit, gasPrice, bridgeOutPayload);
```
Comment (alexanderattar) [inline src/HolographOperator.sol:519]: need to upgrade for this change?

File: src/interface/CrossChainMessageInterface.sol [hunk 3]
```diff
@@ -23,6 +23,6 @@ interface CrossChainMessageInterface {
     uint32 toChain,
     uint256 gasLimit,
     uint256 gasPrice,
-    uint256 payloadLength
+    bytes calldata crossChainPayload
```
Comment (alexanderattar) [inline src/interface/CrossChainMessageInterface.sol:26]: Also just curious why the change?

File: src/module/LayerZeroModule.sol [hunk 4]
```diff
@@ -60,14 +66,16 @@ contract LayerZeroModule is Admin, Initializable, CrossChainMessageInterface, La
       address bridge,
       address interfaces,
       address operator,
+      address optimismGasPriceOracle,
```
Comment (alexanderattar) [inline src/module/LayerZeroModule.sol:69]: This is what you mentioned using Uniswap / Sushiswap for right?

File: deploy/10_layer_zero_module.ts (hunk 1 above)
Comment (ACC01ADE) [reply deploy/10_layer_zero_module.ts:202]: For `minnet` and `testnet` yes, but for `develop` we can deploy this with deployer wallet.

File: src/module/LayerZeroModule.sol [hunk 5]
```diff
@@ -185,6 +193,9 @@ contract LayerZeroModule is Admin, Initializable, CrossChainMessageInterface, La
     require(gasLimit < gasParameters.maxGasLimit, "HOLOGRAPH: gas limit over max");
     (uint256 nativeFee, ) = lz.estimateFees(lzDestChain, address(this), crossChainPayload, false, adapterParams);
     hlgFee = ((gasPrice * gasLimit) * dstPriceRatio) / (10 ** 10);
+    if (toChain == uint32(7) || toChain == uint32(4000000015)) {
```
Comment (alexanderattar) [inline src/module/LayerZeroModule.sol:201]: Maybe adding a comment on these chain ids would be helpful so you don't need to look them up if you haven't memorized all the chain ids 😅

File: src/HolographOperator.sol (hunk 2 above)
Comment (ACC01ADE) [reply src/HolographOperator.sol:519]: Yep, will need to upgrade Holograph Operator contract. Otherwise we will not be able to get a precise L1 Security Fee cost. Since zero bytes and non-zero bytes have different costs associated with them.

File: src/interface/CrossChainMessageInterface.sol (hunk 3 above)
Comment (ACC01ADE) [reply src/interface/CrossChainMessageInterface.sol:26]: To properly calculate the L1 Security Fee, we need entire payload data. Zero bytes and non-zero bytes have different costs associated with them.

File: src/module/LayerZeroModule.sol (hunk 4 above)
Comment (ACC01ADE) [reply src/module/LayerZeroModule.sol:69]: No this is for calculating Optimism L1 Security Fee

File: src/module/LayerZeroModule.sol (hunk 5 above)
Comment (alexanderattar) [reply src/module/LayerZeroModule.sol:201]: especially for the holographId

Comment (ACC01ADE) [reply src/module/LayerZeroModule.sol:201]: Good point, made a commit to address this.

File: src/module/LayerZeroModule.sol [hunk 6]
```diff
@@ -330,6 +364,15 @@ contract LayerZeroModule is Admin, Initializable, CrossChainMessageInterface, La
     }
   }
 
+  /**
+   * @dev Internal function used for getting the Holograph Operator Interface
```
Comment (alexanderattar) [inline src/module/LayerZeroModule.sol]: Is this comment a copy/paste that should be updated?

File: src/module/OVM_GasPriceOracle.sol [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  function getL1Fee(bytes memory _data) external view returns (uint256) {
+    uint256 l1GasUsed = getL1GasUsed(_data);
+    uint256 l1Fee = l1GasUsed * l1BaseFee;
+    uint256 divisor = 10 ** decimals;
+    uint256 unscaled = l1Fee * scalar;
+    uint256 scaled = unscaled / divisor;
+    return scaled;
+  }
+
+  /**
```
Comment (alexanderattar) [inline src/module/OVM_GasPriceOracle.sol:126]: This comment is super helpful

File: src/module/OVM_GasPriceOracle.sol [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+  function getL1GasUsed(bytes memory _data) public view returns (uint256) {
+    uint256 total = 0;
+    for (uint256 i = 0; i < _data.length; i++) {
+      if (_data[i] == 0) {
+        total += 4;
+      } else {
+        total += 16;
+      }
+    }
+    uint256 unsigned = total + overhead;
```
Comment (alexanderattar) [inline src/module/OVM_GasPriceOracle.sol:157]: Is `overhead` defined somewhere?

Comment (alexanderattar) [reply src/module/OVM_GasPriceOracle.sol:157]: never mind. i just noticed it gets set to storage variable in the constructor

File: src/interface/CrossChainMessageInterface.sol (hunk 3 above)
Comment (alexanderattar) [reply src/interface/CrossChainMessageInterface.sol:26]: aha good to know

File: src/module/LayerZeroModule.sol (hunk 5 above)
Comment (alexanderattar) [reply src/module/LayerZeroModule.sol:201]: 👍

File: deploy/10_layer_zero_module.ts (hunk 1 above)
Comment (alexanderattar) [reply deploy/10_layer_zero_module.ts:202]: cool that's what i thought

File: src/module/LayerZeroModule.sol (hunk 6 above)
Comment (ACC01ADE) [reply src/module/LayerZeroModule.sol]: Oops, yep, copy pasta gone bad. Made new commit to fix this.

File: src/module/OVM_GasPriceOracle.sol (hunk 7 above)
Comment (ACC01ADE) [reply src/module/OVM_GasPriceOracle.sol:126]: Copy pasta from original Optimism Gas Price Oracle contract. Kept the commentary since it explains the logic very very well.

Comment (alexanderattar) [reply src/module/OVM_GasPriceOracle.sol:126]: nice

File: src/module/LayerZeroModule.sol (hunk 6 above)
Comment (alexanderattar) [reply src/module/LayerZeroModule.sol]: sweet

**Review by alexanderattar — APPROVED** [review]
It all generally makes sense to me, but we'll probably need to upgrade develop and let it run for testing before we know if we'll need to make any adjustments
